consensus/tendermint/tendermint.go

```diff
@@ -357,43 +291,26 @@ func q(totalVotingPower votingPower) votingPower {
 	return q
 }
 
-func (t *Tendermint[V, H, A]) handleFutureRoundMessage(header MessageHeader[A], addMessage func()) bool {
-	if header.Round > t.state.round {
-		if header.Round-t.state.round > maxFutureRound {
-			return false
-		}
-
-		addMessage()
-
-		if t.uponSkipRound(header.Round) {
-			t.doSkipRound(header.Round)
-		}
-		return false
+func (t *Tendermint[V, H, A]) preprocessMessage(header MessageHeader[A], addMessage func()) bool {
+	var currentRound round
+	isCurrentHeight := header.Height == t.state.height
+	if isCurrentHeight {
+		currentRound = t.state.round
 	}
-	return true
-}
-
-func (t *Tendermint[V, H, A]) handleFutureHeightMessage(header MessageHeader[A], addMessage func()) bool {
-	if header.Height > t.state.height {
-		if header.Height-t.state.height > maxFutureHeight {
-			return false
-		}
-
-		if header.Round > maxFutureRound {
-			return false
-		}
-
-		addMessage()
 
+	switch {
+	case !isCurrentOrNearFuture(header.Height, t.state.height, maxFutureHeight):
+		return false
+	case !isCurrentOrNearFuture(header.Round, currentRound, maxFutureRound):
 		return false
+	default:
+		addMessage()
+		return isCurrentHeight
 	}
-	return true
 }
 
-func (t *Tendermint[V, H, A]) preprocessMessage(header MessageHeader[A], addMessage func()) bool {
-	return header.Height >= t.state.height &&
-		t.handleFutureHeightMessage(header, addMessage) &&
-		t.handleFutureRoundMessage(header, addMessage)
+func isCurrentOrNearFuture[T ~uint | ~int](header, current, maxFuture T) bool {
+	return header >= current && header <= current+maxFuture
 }
```

Comment:
preprocessMessage is a bit hard to understand initially. Can we improve the readability by doing something like the following? 

```suggestion
// preprocessMessage calls `addMessage` if the `msgHeader` is within `maxFutureHeight` of the current height
// and `maxFutureRound` of the current round, otherwise returns false.
func (t *Tendermint[V, H, A]) preprocessMessage(msgHeader MessageHeader[A], addMessage func()) bool {
	var currentRound round
	msgIsCurrentHeight := msgHeader.Height == t.state.height
	if msgIsCurrentHeight {
		currentRound = t.state.round
	}

	switch {
	case !isCurrentOrNearFuture(msgHeader.Height, t.state.height, maxFutureHeight):
		return false
	case !isCurrentOrNearFuture(msgHeader.Round, currentRound, maxFutureRound):
		return false
	default:
		addMessage()
		return msgIsCurrentHeight
	}
}

func isCurrentOrNearFuture[T ~uint | ~int](msgHeader, currentState, maxFuture T) bool {
	return msgHeader >= currentState && msgHeader <= currentState+maxFuture
}
```

Reply:
My suggested changes aren't the cleanest sorry